qiskit/visualization/matplotlib.py

```diff
@@ -414,7 +414,9 @@ def _get_layer_widths(self):
                 self._data[node] = {}
                 self._data[node]["width"] = WID
                 num_ctrl_qubits = 0 if not hasattr(op, "num_ctrl_qubits") else op.num_ctrl_qubits
-                if getattr(op, "_directive", False) or isinstance(op, Measure):
+                if (
+                    getattr(op, "_directive", False) and (not op.label or not self._plot_barriers)
+                ) or isinstance(op, Measure):
```

Comment:
I think the `self._plot_barriers` might be a little overzealous here?  I think this would cause a non-barrier directive to use its name rather than labels if `plot_barriers` is False.

Reply:
Actually, I think `plot_barriers` has evolved at some point in the past to really mean `plot_directive`. All the drawers treat all the directives the same.

Reply:
Oh, that's very confusing.  Still, that's definitely not the domain of this PR, so I think you're right here.